### PR TITLE
feat: File Watcher

### PR DESCRIPTION
**Pull request by eagarwal-notion**
# Motivation
When an agent is modifying the codebase, they may edit many files without calling an LSP tool. This means that the diagnostics will likely be out of date or inaccurate.  
Additionally, even if they call an LSP tool after making changes, we'd need to immediately reparse the file. For larger changes, this can add a lot of delay
# Implementation
- Add a file watcher to watch the entire repository. This filters out gitignored files to avoid watching node modules
- Refactor the didOpen/didSave/didChange/didClose notifications to use dedicated functions
- Leave the existing logic to check if a file is changed during a tool call. In most cases, the file watcher should have already updated it but this will cover edge cases - in particular around tests.
# Misc notes
- This will initialize the project upon a file change, so one may want to disable eagerStartup.

### CODE REVIEW

**Comment by eagarwal-notion** [comment]
* **#25** <a href="https://app.graphite.dev/github/pr/makenotion/lsp-mcp/25?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/makenotion/lsp-mcp/25?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/files.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
+import watcher, { Event } from "@parcel/watcher"
```
Comment (walkerdb) [inline src/files.ts:1]: yess, love the parcel watcher!

File: src/files.test.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import { pathToFileUri } from "./lsp-methods";
+
+describe("File Watcher tests", () => {
+  let watcher: FileWatcher
+  const onAdd = jest.fn<(_: string, __: string) => Promise<void>>()
+  const onChange = jest.fn<(_: string, __: string) => Promise<void>>()
+  const onRemove = jest.fn<(_: string) => Promise<void>>()
+  const testDir = path.resolve("__test__")
+  const target = join(testDir, "test.txt")
+  const target_uri = pathToFileUri(target)
```
[the remark this answers is not in the document]
Comment (eagarwal-notion) [reply src/files.test.ts]: fixed - my python is showing 💀

File: src/FileWatcher.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    private readonly onFileRemoved: (uri: string) => Promise<void>,
+    private readonly onFileCreated: (uri: string, content: string) => Promise<void>,
+  ) {
+    this.events = []
+  }
+  queueEvents(events: Event[]) {
+    for (const fs_event of events) {
+      if (!this.extensions.some(ext => fs_event.path.endsWith(ext))) {
+        return
+      }
```
Comment (walkerdb) [inline src/FileWatcher.ts:43]: doesn't this mean the entire event will be dismissed if even just one file in the event list doesn't match the extension list?

File: src/lsp.ts [hunk 4]
```diff
@@ -368,6 +404,7 @@ export class LspClientImpl implements LspClient {
   }
   async dispose() {
     try {
+      this.fileWatcher.dispose()
```
[the remark this answers is not in the document]
Comment (eagarwal-notion) [reply src/lsp.ts:407]: Fixed

File: src/FileWatcher.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    private readonly logger: Logger,
+    private readonly onFileChanged: (uri: string, content: string) => Promise<void>,
+    private readonly onFileRemoved: (uri: string) => Promise<void>,
+    private readonly onFileCreated: (uri: string, content: string) => Promise<void>,
+  ) {
+    this.events = []
+  }
+  queueEvents(events: Event[]) {
+    for (const fs_event of events) {
+      if (!this.extensions.some(ext => fs_event.path.endsWith(ext))) {
```
Comment (walkerdb) [inline src/FileWatcher.ts:41]: probably unimportant perf nit: if `extensions` only exists to check membership, should it be a `Set` instead?

Comment (walkerdb) [reply src/FileWatcher.ts:41]: kk makes sense to me!
